File: June-LeetCoding-Challenge-2020/Week3/130-surrounded-region.cpp
// Solution using DFS
// TC: O(n^2) has to go through all 
// SC: O(n)
// Runtime: 24 ms, faster than 96.98%
// Memory Usage: 10 MB, less than 96.37% 
class Solution {
public:
    int row, col;
    
    void DFS(vector<vector<char>> &board, int r, int c){
        
        if(r < 0 || r > row-1 || c < 0 || c > col-1 || board[r][c] != 'O')
            return;
        
        board[r][c] = 'B';
        
        DFS(board, r-1, c);
        DFS(board, r+1, c);
        DFS(board, r, c-1);
        DFS(board, r, c+1);
    }
    
    void solve(vector<vector<char>>& board) {
        
        if(!board.size())
            return;
        
        row = board.size();
        col = board[0].size();
        
        for(int i = 0; i < row; i++){
            DFS(board, i, 0);
            DFS(board, i, col-1);
        }
        
        for(int i = 0; i < col; i++){
            DFS(board, 0, i);
            DFS(board, row-1, i);
        }
        
        for(int i = 0; i < row; i++)
            for(int j = 0; j < col; j++){
                if(board[i][j] == 'B'){
                    board[i][j] = 'O';
                }else if(board[i][j] == 'O'){
                    board[i][j] = 'X';
                }
            }
        
    }
};

File: June-LeetCoding-Challenge-2020/Week3/130-surrounded-region.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        if not board:
            return 
        height = len(board)
        width = len(board[0])
        
        def checkO(r,c):
            if 0<=r<height and 0<=c<width and board[r][c] == 'O': # within in the boundary
                board[r][c] = 'C'
                checkO(r-1,c)
                checkO(r+1,c)
                checkO(r,c-1)
                checkO(r,c+1)

        for r in [0,height-1]: # first row, last row
            for c in range(width):
                checkO(r,c)
        for c in [0,width-1]: # first col, last col
            for r in range(height):
                checkO(r,c)
        for r in range(height):
            for c in range(width):
                if board[r][c] == 'C':
                    board[r][c] = 'O'            
                elif board[r][c] == 'O':
                    board[r][c] = 'X'
    # TC: O(H*W)
    # SC: O(N), n = h + w if anything contains O on the boundary
    # Runtime: 148 ms, faster than 55.81% 
    # Memory Usage: 15.6 MB, less than 34.04%  